### DISCUSSION TITLE
Missing data when refining a union

### DISCUSSION THREAD

**Discussion by bstro** [colinhacks/zod#3509] [Q&A]
Given the following simple schema with attached `refine` logic:
```
const exampleUnion = z
  .union([z.object({ foo: z.string() }), z.object({ bar: z.string() })])
  .refine((data) => {
    if ('foo' in data && 'bar' in data) {
      return false;
    }
    return true;
  });
```

In this hypothetical scenario that mirrors some acceptance criteria, I need to surface an error to the user if they have supplied both `foo` and `bar`. However, when I try to parse an object like `exampleUnion.parse({foo: 'hello world', bar: 'test' })`, the resulting output keeps the first member of the union and strips out the rest:
```
{
  "foo": "hello world"
}
```

Any ideas how I can detect whether the end user has passed both `foo` and `bar`?

### Reproduction
https://stackblitz.com/edit/vitejs-vite-3vwqsq?file=package.json,src%2Fmain.ts&terminal=dev

**Comment by marilari88** [comment]
hi @bstro, you can leverage the z.undefined() primitive inside the union.
```typescript
z.union(
  [
    z.object({ foo: z.string(), bar: z.undefined() }),
    z.object({ bar: z.string(), foo: z.undefined() })
  ]
)
```

https://zod-playground.vercel.app?appdata=N4IgzgxgFgpgtgQxALhALwHQFcB2BLAexwAoAdHAAgoG1yqrMCAjAKxggBdjgKAzAgsgqYwHAE54cAc2IBKADQUmCMUMy4AJjF6SYGuRQC%2BCuvUat2XHstXCMoidLmL%2Bgu5u279so7NMBdclkQeRAANwQAGywYMBRqEB5XIQByWEjIggoAdwIxSI0UxRtUjliOFKMQxL4BVPTMnLyCysNq4BKKFLLRVpB-ULCYMTBCHBQQAGYMACZpgA4QQyA
